File: core/engines/python_engine/utils/task_runner.py
"""
Task runner with timeout support and result collection.
"""

import asyncio
from typing import Any, Callable, List, Optional
from dataclasses import dataclass
import time
from .errors import TimeoutError as EngineTimeoutError

@dataclass
class TaskResult:
    """Result of a task execution."""
    name: str
    success: bool
    duration_ms: float
    message: str
    output: Any = None

class TaskRunner:
    """Executes tasks with timeout support and collects results."""
    
    def __init__(self, max_concurrent: int = 4):
        self.max_concurrent = max_concurrent
        self.results: List[TaskResult] = []
    
    def run_sync(self, name: str, func: Callable, timeout_seconds: Optional[float] = None, *args: Any, **kwargs: Any) -> TaskResult:
        """
        Run a synchronous function with optional timeout.
        
        :param name: Task name
        :param func: Callable to execute
        :param timeout_seconds: Optional timeout in seconds
        :param args: Positional arguments for func
        :param kwargs: Keyword arguments for func
        :return: TaskResult
        """
        start = time.time()
        try:
            if timeout_seconds:
                result_container = [None]
                exception_container = [None]
                
                def wrapper():
                    try:
                        result_container[0] = func(*args, **kwargs)
                    except Exception as e:
                        exception_container[0] = e
                
                import threading
                thread = threading.Thread(target=wrapper)
                thread.daemon = True
                thread.start()
                thread.join(timeout=timeout_seconds)
                
                if thread.is_alive():
                    raise EngineTimeoutError(f"Task '{name}' exceeded timeout of {timeout_seconds}s")
                
                if exception_container[0]:
                    raise exception_container[0]
                
                output = result_container[0]
            else:
                output = func(*args, **kwargs)
            
            duration_ms = (time.time() - start) * 1000
            result = TaskResult(
                name=name,
                success=True,
                duration_ms=duration_ms,
                message="Task completed successfully",
                output=output
            )
        except Exception as e:
            duration_ms = (time.time() - start) * 1000
            result = TaskResult(
                name=name,
                success=False,
                duration_ms=duration_ms,
                message=str(e),
                output=None
            )
        
        self.results.append(result)
        return result
    
    async def run_async(self, name: str, coro: Callable, timeout_seconds: Optional[float] = None, *args: Any, **kwargs: Any) -> TaskResult:
        """
        Run an async coroutine with optional timeout.
        
        :param name: Task name
        :param coro: Async callable
        :param timeout_seconds: Optional timeout in seconds
        :param args: Positional arguments
        :param kwargs: Keyword arguments
        :return: TaskResult
        """
        start = time.time()
        try:
            if timeout_seconds:
                output = await asyncio.wait_for(
                    coro(*args, **kwargs),
                    timeout=timeout_seconds
                )
            else:
                output = await coro(*args, **kwargs)
            
            duration_ms = (time.time() - start) * 1000
            result = TaskResult(
                name=name,
                success=True,
                duration_ms=duration_ms,
                message="Task completed successfully",
                output=output
            )
        except asyncio.TimeoutError:
            duration_ms = (time.time() - start) * 1000
            result = TaskResult(
                name=name,
                success=False,
                duration_ms=duration_ms,
                message=f"Task exceeded timeout of {timeout_seconds}s",
                output=None
            )
        except Exception as e:
            duration_ms = (time.time() - start) * 1000
            result = TaskResult(
                name=name,
                success=False,
                duration_ms=duration_ms,
                message=str(e),
                output=None
            )
        
        self.results.append(result)
        return result
    
    def get_results(self) -> List[TaskResult]:
        """Get all collected results."""
        return self.results.copy()
    
    def clear_results(self) -> None:
        """Clear all results."""
        self.results.clear()
    
    def get_summary(self) -> str:
        """Get summary of execution."""
        total = len(self.results)
        succeeded = sum(1 for r in self.results if r.success)
        failed = total - succeeded
        total_time = sum(r.duration_ms for r in self.results)
        
        return (
            f"TaskRunner Summary: {total} total, "
            f"{succeeded} succeeded, {failed} failed, "
            f"{total_time:.2f}ms total"
        )

File: core/engines/rust_engine/src/core/task_runner.rs
// src/task_runner.rs
use crate::utils::error::{EngineError, Result};
use std::time::Duration;
use tokio::time::timeout;
use std::sync::Arc;

/// Task execution result
#[derive(Debug, Clone)]
pub struct TaskResult {
    pub name: String,
    pub success: bool,
    pub duration_ms: u128,
    pub message: String,
}

/// Task metadata and execution context
pub struct Task {
    pub name: String,
    pub timeout: Option<Duration>,
    pub priority: u8,
}

impl Task {
    pub fn new(name: String, priority: u8) -> Self {
        Task {
            name,
            timeout: None,
            priority,
        }
    }

    pub fn with_timeout(mut self, timeout: Duration) -> Self {
        self.timeout = Some(timeout);
        self
    }
}

/// TaskRunner manages execution of async tasks with timeouts and resource limits
pub struct TaskRunner {
    max_concurrent: usize,
    results: Arc<tokio::sync::Mutex<Vec<TaskResult>>>,
}

impl TaskRunner {
    pub fn new(max_concurrent: usize) -> Self {
        TaskRunner {
            max_concurrent,
            results: Arc::new(tokio::sync::Mutex::new(Vec::new())),
        }
    }

    /// Execute an async function with optional timeout
    pub async fn run<F, Fut>(&self, task: Task, f: F) -> Result<TaskResult>
    where
        F: FnOnce() -> Fut,
        Fut: std::future::Future<Output = Result<String>>,
    {
        let start = std::time::Instant::now();

        let result = if let Some(timeout_duration) = task.timeout {
            match timeout(timeout_duration, f()).await {
                Ok(Ok(msg)) => Ok(msg),
                Ok(Err(e)) => Err(e),
                Err(_) => Err(EngineError::TimeoutError(format!(
                    "Task '{}' exceeded timeout of {:?}",
                    task.name, timeout_duration
                ))),
            }
        } else {
            f().await
        };

        let duration_ms = start.elapsed().as_millis();
        let (success, message) = match result {
            Ok(msg) => (true, msg),
            Err(e) => (false, e.to_string()),
        };

        let task_result = TaskResult {
            name: task.name.clone(),
            success,
            duration_ms,
            message,
        };

        self.results.lock().await.push(task_result.clone());
        Ok(task_result)
    }

    /// Get all collected results
    pub async fn get_results(&self) -> Vec<TaskResult> {
        self.results.lock().await.clone()
    }

    /// Get the maximum concurrent tasks
    pub fn max_concurrent(&self) -> usize {
        self.max_concurrent
    }

    /// Get summary of results
    pub async fn get_summary(&self) -> String {
        let results = self.results.lock().await;
        let total = results.len();
        let succeeded = results.iter().filter(|r| r.success).count();
        let failed = total - succeeded;
        let total_time: u128 = results.iter().map(|r| r.duration_ms).sum();

        format!(
            "TaskRunner Summary: {} total, {} succeeded, {} failed, {} ms total",
            total, succeeded, failed, total_time
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_task_execution() {
        let runner = TaskRunner::new(4);
        let task = Task::new("test_task".to_string(), 0);
        let result = runner
            .run(task, || async { Ok("success".to_string()) })
            .await;

        assert!(result.is_ok());
        let r = result.unwrap();
        assert!(r.success);
        assert_eq!(r.message, "success");
    }

    #[tokio::test]
    async fn test_task_timeout() {
        let runner = TaskRunner::new(4);
        let task = Task::new("timeout_task".to_string(), 0)
            .with_timeout(Duration::from_millis(100));

        let result = runner
            .run(task, || async {
                tokio::time::sleep(Duration::from_secs(1)).await;
                Ok("should not reach".to_string())
            })
            .await;

        assert!(result.is_ok());
        let r = result.unwrap();
        assert!(!r.success); // Task timed out
    }
}
